### PR TITLE
NOISSUE - Check For Policy Action Before Adding Policies

### PR DESCRIPTION
**Pull request by rodneyosodo**
### What does this do?
Check for policy actions before adding policy. If the current users has `g_add` and `c_list` policies on a group, they can only add members to that group with either `g_add` or `c_list` actions. If they try to add a member with `c_update` it should fail unless they are the owner of that group.

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
When adding members to a group, add members with equal policy like yours or subset of your policies.

### Have you included tests for your changes?
Yes

### Did you document any new/modified functionality?
No

### Notes
How to reproduce this while testing is
1. Create 3 different users `a`, `b` and `c`.
2. Create a group with user `a`
3. Add user `b` to the group with only `g_add` actions while using the token of user `a`
4. Try to add user `c` to the group with `c_update` actions while using the token of user `b`

### CODE REVIEW

File: things/policies/service.go [hunk 1]
```diff
@@ -127,8 +127,21 @@ func (svc service) AddPolicy(ctx context.Context, token string, p Policy) (Polic
 	}
 
 	// If the client has `g_add` action on the object or is the owner of the object, add the policy
-	ar := AccessRequest{Subject: userID, Object: p.Object, Action: "g_add"}
-	if _, err := svc.policies.EvaluateGroupAccess(ctx, ar); err == nil {
+	areq := AccessRequest{Subject: userID, Object: p.Object, Action: "g_add"}
+	if _, err := svc.policies.EvaluateGroupAccess(ctx, areq); err == nil {
+		page := Page{Subject: userID, Object: p.Object, Offset: 0, Limit: 1}
+		pol, err := svc.policies.Retrieve(ctx, page)
```
Comment (dborovcanin) [inline things/policies/service.go]: Is this fetch necessary and why don't we use the policy from `EvaluateGroupAccess`?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: things/policies/service.go (hunk 1 above)
Comment (rodneyosodo) [reply things/policies/service.go]: Because it was returning only `subject`. I have changed it to return `SOA`

File: things/policies/policies.go [hunk 2]
```diff
@@ -135,3 +136,22 @@ func ValidateAction(act string) bool {
...
+func checkActions(currentActions, incomingActions []string) error {
+	for _, action := range incomingActions {
+		if !contains(currentActions, action) {
+			return errors.ErrAuthorization
+		}
+	}
+	return nil
+}
+
+func contains(actions []string, action string) bool {
```
Comment (SammyOina) [inline things/policies/policies.go]: you can use golang.org/x/exp/slices with the slices.Contains() method

File: users/policies/policies.go [hunk 3]
```diff
@@ -208,3 +209,22 @@ func removeDuplicates(slice []string) []string {
...
+func checkActions(currentActions, incomingActions []string) error {
+	for _, action := range incomingActions {
+		if !contains(currentActions, action) {
+			return errors.ErrAuthorization
+		}
+	}
+	return nil
+}
+
+func contains(actions []string, action string) bool {
```
Comment (SammyOina) [inline users/policies/policies.go]: same here

**Review by SammyOina — CHANGES_REQUESTED** [review]

**Review by SammyOina — APPROVED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Pull request merged**